### PR TITLE
Added Binary Heap Data Structure

### LINKED ISSUE #146: Add B-heap data structure
#### References/Other comments
[1] https://en.wikipedia.org/wiki/B-heap

### PR DESCRIPTION
**Pull request by sarthakforwet**
#### References to other Issues or PRs or Relevant literature
Fixes #146 

#### Brief description of what is fixed or changed
Added B-heap DS in both Min and Max Mode.

#### Other comments
In my opinion this module has been completed but if any method that others may feel should be there , are requested to be suggested to me .

### CODE REVIEW

File: build/lib/pydatastructs/miscellaneous_data_structures/linked_stacks.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+from pydatastructs import SinglyLinkedList
+
+__all__ = ["Linked_Stacks"]
+class Linked_Stacks:
+    
+    # A class to implement Stacks Built over Singly Linked Lists
+
```
Comment (HarsheetKakar) [inline build/lib/pydatastructs/miscellaneous_data_structures/linked_stacks.py]: why is it implemented in both stack.py and linked_stacks.py?

File: build/lib/pydatastructs/miscellaneous_data_structures/linked_stacks.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        return self
+
+    def pop(self):
+        if self.top is not None:
+            self.data = self.top.data
+            self.top = self.top.next
+        return self.data
+    
+    def peek(self):
+        return self.top.data
```
Comment (HarsheetKakar) [inline build/lib/pydatastructs/miscellaneous_data_structures/linked_stacks.py]: please add newline at the end of the file

File: build/lib/pydatastructs/miscellaneous_data_structures/stack.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    .. [1] https://en.wikipedia.org/wiki/Stack_(abstract_data_type)
+    """
+
+    def __new__(cls, implementation='array', **kwargs):
+        if implementation == 'array':
+            return ArrayStack(
+                kwargs.get('items', None),
+                kwargs.get('dtype', int))
+
+        elif implementation == "ll":
```
Comment (HarsheetKakar) [inline build/lib/pydatastructs/miscellaneous_data_structures/stack.py]: I think in the queue module implementation is written as linkedlist

File: build/lib/pydatastructs/miscellaneous_data_structures/stack.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+            self.top = self.dll.head
+        self.top = self.top.next
+
+    def pop(self):
+        if not self.is_empty:
+            self.data = self.top.data
+            self.top = self.top.prev
+            return self.data
+        else:
+            raise ValueError
```
Comment (HarsheetKakar) [inline build/lib/pydatastructs/miscellaneous_data_structures/stack.py]: please add test for these functions.

File: pydatastructs/miscellaneous_data_structures/priority_queue.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
+from heapq import heapify,heappush,heappop
```
Comment (HarsheetKakar) [inline pydatastructs/miscellaneous_data_structures/priority_queue.py]: IMO, we should use `pydatastructs.trees.heaps`

File: pydatastructs/miscellaneous_data_structures/queue.py [hunk 6]
```diff
@@ -180,3 +179,56 @@ def __len__(self):
 
     def __str__(self):
         return str(self.queue)
+class PriorityQueue(Queue):
```
Comment (HarsheetKakar) [inline pydatastructs/miscellaneous_data_structures/queue.py]: is there a difference between the implementation in priority_queue.py and this one?

**Comment by czgdp1807** [comment]
B-Heap and Binary heaps are not the same. Closing this as the feature proposed is already there in the `master`.
Thanks for the contributions.

File: pydatastructs/miscellaneous_data_structures/priority_queue.py (hunk 5 above)
Comment (sarthakforwet) [reply pydatastructs/miscellaneous_data_structures/priority_queue.py]: Is it necessary to do so? 
Can't we use some external libraries as `heapq` is an inbuilt library in python?

**Pull request closed without merging**